### PR TITLE
Split up babel-core's File class and add Flowtype annotations

### PR DESCRIPTION
**Pull request by loganfsmyth**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues             | 
| Patch: Bug Fix?          | N
| Major: Breaking Change?  | N, unless you're using undocumented methods/props on `File`.
| Minor: New Feature?      | 
| Tests Added/Pass?        | 
| Spec Compliancy?         | 
| License                  | MIT
| Doc PR                   | 
| Any Dependency Changes?  | 

FYI, may be easiest to review by commit. Mostly refactoring. I've removed the parts of the `File` class that weren't part of the user-facing API and moved then to be their own files.

Actual features in here:
* Added `browser` files for the 2 `transformFile*` files that actually rely on `fs` so Webpack will complain less.
* Flowtype on every file in `babel-core` now except `build-external-helpers`.
* We now get useful code snippets in exceptions thrown by `babel-template` 
* It's a bit easier to follow the transform process now that it's not a chain of class methods.

### CODE REVIEW

File: packages/babel-traverse/src/hub.js [hunk 1]
```diff
@@ -1,6 +1,5 @@
 export default class Hub {
-  constructor(file, options) {
```
Comment (hzoo) [inline packages/babel-traverse/src/hub.js]: is it even worth having this file if it's just a file?

Comment (loganfsmyth) [reply packages/babel-traverse/src/hub.js]: Yeah we can revisit, I just didn't want to change any of the actual transform infrastructure.

File: packages/babel-core/src/index.js [hunk 2]
```diff
@@ -1,39 +1,42 @@
-export File from "./transformation/file";
```
Comment (thejameskyle) [inline packages/babel-core/src/index.js]: Can we continue to export `File`? I need it in a number of places to create nice error messages

Comment (jridgewell) [reply packages/babel-core/src/index.js]: @loganfsmyth

**Pull request merged**